### PR TITLE
Implement JIT helper for resolve constantDynamic

### PR DESCRIPTION
**Pull request by fengxue-IS**
Close: #1277

Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

### CODE REVIEW

File: runtime/codert_vm/cnathelp.cpp [hunk 1]
```diff
@@ -1745,6 +1745,37 @@ old_slow_jitResolveInvokeDynamic(J9VMThread *currentThread)
...
+	OLD_SLOW_ONLY_JIT_HELPER_PROLOGUE(3);
+	DECLARE_JIT_PARM(J9ConstantPool*, ramConstantPool, 1);
+	DECLARE_JIT_INT_PARM(cpIndex, 2);
+	DECLARE_JIT_PARM(void*, jitEIP, 3);
+	void *addr = NULL;
+	J9RAMConstantDynamicRef *ramCPEntry = (J9RAMConstantDynamicRef*)ramConstantPool + cpIndex;
+	j9object_t *constantDynamicValue = &ramCPEntry->value;
+
+	/* Check if the value is resolved, Void.Class exception represents a valid null reference */
+	if (NULL == *constantDynamicValue && ramCPEntry->exception != currentThread->javaVM->voidReflectClass->classObject) {
```
Comment (gacholio) [inline runtime/codert_vm/cnathelp.cpp]: Need some brackets around the two comparisions.

File: runtime/codert_vm/cnathelp.cpp [hunk 2]
```diff
@@ -1745,6 +1745,37 @@ old_slow_jitResolveInvokeDynamic(J9VMThread *currentThread)
...
+		/* If entry resolved to an exception previously, same exception will be set by resolution code */
+		buildJITResolveFrameWithPC(currentThread, J9_SSF_JIT_RESOLVE_DATA, parmCount, true, 0, jitEIP);
+		constantDynamicValue = &(currentThread->javaVM->internalVMFunctions->resolveConstantDynamic(
+																				currentThread,
```
Comment (gacholio) [inline runtime/codert_vm/cnathelp.cpp]: Looks like some overindenting here (unless it's just github).

File: runtime/codert_vm/cnathelp.cpp [hunk 3]
```diff
@@ -1745,6 +1745,37 @@ old_slow_jitResolveInvokeDynamic(J9VMThread *currentThread)
...
+	DECLARE_JIT_PARM(void*, jitEIP, 3);
+	void *addr = NULL;
+	J9RAMConstantDynamicRef *ramCPEntry = (J9RAMConstantDynamicRef*)ramConstantPool + cpIndex;
+	j9object_t *constantDynamicValue = &ramCPEntry->value;
+
+	/* Check if the value is resolved, Void.Class exception represents a valid null reference */
+	if (NULL == *constantDynamicValue && ramCPEntry->exception != currentThread->javaVM->voidReflectClass->classObject) {
+		/* If entry resolved to an exception previously, same exception will be set by resolution code */
+		buildJITResolveFrameWithPC(currentThread, J9_SSF_JIT_RESOLVE_DATA, parmCount, true, 0, jitEIP);
+		constantDynamicValue = &(currentThread->javaVM->internalVMFunctions->resolveConstantDynamic(
```
Comment (gacholio) [inline runtime/codert_vm/cnathelp.cpp]: This likely won't even compile, and certainly doesn't do what you want. Just refetch the pointer after the addr NULL check.

**Comment by fengxue-IS** [comment]
Changed the resolve path to use a `retry` similar to `jitResolveInvokeDynamic`, added brackets to the if check

**Comment by gacholio** [comment]
jenkins compile xlinux jdk8

**Comment by gacholio** [comment]
jenkins compile plinux jdk8

**Comment by gacholio** [comment]
jenkins compile zlinux jdk8

File: runtime/codert_vm/cnathelp.cpp [hunk 4]
```diff
@@ -1745,6 +1745,35 @@ old_slow_jitResolveInvokeDynamic(J9VMThread *currentThread)
...
 }
 
+void* J9FASTCALL
+old_slow_jitResolveConstantDynamic(J9VMThread *currentThread)
+{
+	OLD_SLOW_ONLY_JIT_HELPER_PROLOGUE(3);
+	DECLARE_JIT_PARM(J9ConstantPool*, ramConstantPool, 1);
+	DECLARE_JIT_INT_PARM(cpIndex, 2);
+	DECLARE_JIT_PARM(void*, jitEIP, 3);
+	void *addr = NULL;
```
Comment (gacholio) [inline runtime/codert_vm/cnathelp.cpp]: Declare vm local here to avoid the multiple fetches below.

**Comment by gacholio** [comment]
You also need to add the new function to jilconsts.c

**Comment by fengxue-IS** [comment]
fixed, thanks for pointing out

**Pull request merged**